kernel/term.ml

```diff
@@ -265,6 +265,34 @@ let decompose_lam_assum =
   in
   lamdec_rec Context.Rel.empty
 
+(** Simulaneous decompose_lam_assum / decompose_prod_assum: transforms
+    a pair of lambda term λ (x1:T1)..(xn:Tn). t and type Π
+    (x1:T1)..(xn:Tn). T into the triple ([(xn,Tn);...;(x1,T1)],t,T),
+    where [t] is not lambda, [T] is not a prod, and the first element
+    is the abstracted context *)
+let decompose_lam_prod_assum c ty =
+  let open Context.Rel.Declaration in
+  let rec aux ctx c ty =
+    match kind c, kind ty with
+    | LetIn (x, b, t, c), LetIn (x', b', t', ty)
+      when eq_annot Name.equal x x' && Constr.equal b b' && Constr.equal t t' ->
+      let ctx' = Context.Rel.add (LocalDef (x, b', t')) ctx in
+      aux ctx' c ty
+    | _, LetIn (x', b', t', ty) ->
+      let ctx' = Context.Rel.add (LocalDef (x', b', t')) ctx in
+      aux ctx' (lift 1 c) ty
+    | LetIn (x, b, t, c), _ ->
+      let ctx' = Context.Rel.add (LocalDef (x, b, t)) ctx in
+      aux ctx' c (lift 1 ty)
+    | Lambda (x, b, t), Prod (x', b', t')
+      when eq_annot Name.equal x x' && Constr.equal b b' ->
```

Comment:
You changed `(* By invariant, must be convertible *)` to a syntactic equality check, why?

Reply:
I tried to make the API stronger; it worked on the test suite. On the other hand `equations` fails, I dunno if that's related to this tho as it uses the other API.